### ISSUE TITLE
[Question] Creating and debugging core dumps on Linux

### ISSUE THREAD

**Issue by richardszalay** [dotnet/coreclr#13937]
Hi guys,

I'm attempting to "port" Tess Ferrandez's debugging labs over to .NET Core, including working through them on the various platforms supported by .NET Core.

On Windows the story was pretty easy (adplus has changed, but it's still a thing), but I'm having a lot more problems on Linux (Ubuntu Server 16.04).

Essentially I can't seem to successfully debug a coredump created by anything other than `createdump` that ships with the .NET Core SDK (2.0, via apt-get). 

| |Size|lldb-3.6|lldb-3.7|
| ----- | ----- | ---- | --- |
|gcore parent|3GB|Process must be launched|lldb crashed (Segmentation fault)|
|gcore child|11GB|Process must be launched|lldb crashed (Segmentation fault)|
|coredump parent|147MB|Success|Success|
|coredump child|163MB|Success|Success|
|appport|29MB|Unable to find process plug-in for core file|Unable to find process plug-in for core file|

("parent" here is the "dotnet run" process, child is "dotnet exec", appport crash is caused by throwing an exception in a finalizer thread)

Can anyone tell me what I might be doing wrong?

Rich

**Comment by janvorli** [comment]
@richardszalay regarding the lldb-3.7 crashes, are these happening when you load the dump or when you try to load the libsosplugin.so?

**Comment by richardszalay** [comment]
@janvorli When I load the dump

**Comment by richardszalay** [comment]
Oh, and FWIW, "Success" is defined being able perform the following commands successfully:

```
lldb-3.x -c /var/temp/core /usr/share/dotnet/dotnet
thread list
plugin load /usr/share/dotnet/shared/Microsoft.NETCore.App/2.0.0.0/libsosplugin.so
clrthreads
```

**Comment by janvorli** [comment]
Actually, lldb 3.7 is known to have problems with dump loading. Moreover, the libsosplugin.so that's in the dotnet core package is for lldb 3.6 only. Unfortunately, the plugin needs to be built against the same version of lldb libraries as the lldb that will use it. While it seems to load ok even to different versions, it doesn't work and often crashes badly when you try to run some SOS commands.
You'd need to build libsosplugin.so for the target lldb yourself from coreclr sources.

**Comment by richardszalay** [comment]
@janvorli Thanks for the heads up. Any thoughts as to the "Process must be launched" I receive when loading the gcore dump from lldb-3.6?

**Comment by janvorli** [comment]
Please see this work in progress PR for more details:
https://github.com/mikem8361/coreclr/blob/5c22cb85c7cc9173f2fb783bf24c0cbbb6096c89/Documentation/building/debugging-instructions.md

**Comment by richardszalay** [comment]
Amazing, looks like a great source of information.

**Comment by albigi** [comment]
Guys, I'm actually getting this very same behavior, plus one more oddity.

If I run a dotnet 2.0.0 app on a Ubuntu 14.04, I dump the process using createdump -u (full core dump), then I try to load it in lldb3.6 and it fails to load the dac:

_Failed to load data access DLL, 0x80004005
Can not load or initialize libmscordaccore.so. The target runtime may not be initialized.
ClrStack  failed_

If I run a dotnet 1.1.4 app, createdump is not available and lldb-3.6 also fails to create a coredump (no ObjectFile plugins were able to save a core for this process).
the core dump created with gcore (gdb 7.7.1) gives the "Process must be launched" issue.

long story short:

- it seems dotnet 2.0.0 wants lldb 3.8 or later to properly resolve the process info but those versions of lldb are not compatible with the dbg plugin that comes with the SDK.

- it seems lldb-3.6 is not compatible with the core format of gcore (at least the 7.7.1 version)

any way out of this? would compiling the whole 2.0.0 SDK on a Ubuntu 17 with LLDB-4 possibly work to analyze dumps from a 14.04 distro?

**Comment by janvorli** [comment]
@albigi the easiest way is to install llvm 3.9 (clang, lldb, lldb-dev) on Ubuntu 14.04, build coreclr repo and then use lldb-3.9 with the libsosplugin.so the build has produced. To install llvm-3.9, you first need to add the repo where it is located:
```
echo "deb http://apt.llvm.org/trusty/ llvm-toolchain-trusty-3.9 main" | sudo tee /etc/apt/sources.list.d/llvm.list
wget -O - http://llvm.org/apt/llvm-snapshot.gpg.key | sudo apt-key add -
sudo apt-get update
```
Then install llvm-3.9, clang-3.9, lldb-3.9 and liblldb-3.9-dev and build coreclr repo, passing the build `clang3.9` option. This option is not strictly necessary, but it is better to match the co
